### PR TITLE
ARROW-10747: [Rust]: CSV reader optimization

### PR DESCRIPTION
**Pull request by Dandandan**
this PR makes CSV reading (quite a bit) faster by reusing allocations, and doing things a bit more manually.
It removes usage of BufReader, which is done in rust-csv already and causes overhead. 

The nytaxi (entire job, with reading 1 year csv) benchmark speeds up from ~4500ms to ~1900ms.
Loading the line item csv in memory for the tpch benchmark for goes from ~9800ms -> ~6000 ms.

I think a further optimization would be to stop using the `StringRecords` altogether (e.g. by using the underlying https://docs.rs/csv-core/0.1.10/csv_core/ library instead) but that could be a next step.

FYI @alamb @nevi-me @jorgecarleitao

### CODE REVIEW

**Comment by Dandandan** [comment]
I found some further opportunities for optimizing by also reusing the stringrecord items, for another speed up.

**Comment by Dandandan** [comment]
Is ready for review now.

**Review by jorgecarleitao — APPROVED** [review]
Nice work, @Dandandan , really cool speedup for such an important op.

**Review by nevi-me — APPROVED** [review]
LGTM

File: rust/benchmarks/src/bin/nyctaxi.rs
Comment (Dandandan) [reply rust/benchmarks/src/bin/nyctaxi.rs]: Great, addressed!

**Comment by andygrove** [comment]
@Dandandan I'm not sure why the Windows build failed and I assume it is unrelated but the logs are not available. Could you push an empty commit to trigger CI again?

**Comment by Dandandan** [comment]
> @Dandandan I'm not sure why the Windows build failed and I assume it is unrelated but the logs are not available. Could you push an empty commit to trigger CI again?

Just did. Let's see what happens!

**Comment by nevi-me** [comment]
> Could you push an empty commit to trigger CI again?

We can rerun failed CI jobs from the UI, which is often better as it doesn't trigger AppVeyor and Travis CI

**Comment by Dandandan** [comment]
Now some other jobs failed. Maybe we can rerun those?

**Comment by nevi-me** [comment]
> Now some other jobs failed. Maybe we can rerun those?

CI seems to be misbehaving, it's not letting me cancel the workflow, even though the tests have failed. I'll leave this tab open, and retry the Rust jobs in about an hour. I'll merge this after CI passes

**Comment by Dandandan** [comment]
I also removed the now unused buffered iterator as it is unused by now, and I think will not lead to efficient code in general.

File: rust/benchmarks/src/bin/nyctaxi.rs
Comment (alamb) [inline rust/benchmarks/src/bin/nyctaxi.rs:95]: 👍

**Review by alamb — COMMENTED** [review]
Sorry I didn't get a chance to review this carefully before merge. It looks nice to me. Nice work @Dandandan

**Pull request closed without merging**
